### PR TITLE
test: Build/run with crypto and/or crc

### PR DESCRIPTION
**Pull request by howjmay**
The current test has a lack of tests with crypto flag enabled which cause the tests not cover the total implementation.

### CODE REVIEW

File: .github/workflows/github_actions.yml [hunk 1]
```diff
@@ -69,11 +70,17 @@ jobs:
...
             CXX: ${{ matrix.cxx_compiler }}
+            FEATURE: ${{matrix.feature}}
           install: |
             apt-get update -q -y
             apt-get install -q -y "${{ matrix.cxx_compiler }}" make
             apt-get install -q -y gcc
-          run: make check
+          run: |
+            if [ $FEATURE == "crypto" ]; then
+              make check-crypto
```
Comment (jserv) [inline .github/workflows/github_actions.yml:80]: I didn't see any rules associated with `check-crypto`.

File: Makefile [hunk 2]
```diff
@@ -63,6 +63,12 @@ ifeq ($(processor),$(filter $(processor),aarch64 arm64 arm armv7l))
 endif
 	$(EXEC_WRAPPER) $^
 
+check-crypto: tests/main
```
Comment (jserv) [inline Makefile:66]: How about accepting a variable `FEATURES` which can be recognized as the corresponding flags?
See https://github.com/sysprog21/rv32emu/blob/master/Makefile#L69

Sample usage:
```shell
make FEATURE=crypto clean all
make FEATURE=crypto,crc clean all
```

**Review by jserv — CHANGES_REQUESTED** [review]
Split the changes against `sse2neon.h` to another pull request.

File: Makefile (hunk 2 above)
Comment (howjmay) [reply Makefile:66]: it seems we need to exclude `armv7` from testing crypto flag?

Comment (jserv) [reply Makefile:66]: ARMv8 introduced hardware acceleration for cryptography. From the perspective of Armv8, it is possible to use the cores with 32-bit mode only. That is, A32, which allows the use of cryptographic extension as well.

To make it clear, we should consider the following targets:
* Armv7-A: 32-bit. NEON is the necessary feature for SSE2NEON.
* Armv8-A: 64-bit and/or 32-bit. It comes with several extensions such as cryptography and CRC.

See also: https://github.com/ThomasKaiser/sbc-bench/blob/master/results/ARMv8-Crypto-Extensions.md

File: Makefile [hunk 3]
```diff
@@ -63,6 +69,12 @@ ifeq ($(processor),$(filter $(processor),aarch64 arm64 arm armv7l))
...
+ifeq ($(processor),$(filter $(processor),aarch64 arm64 arm armv7l))
+	$(CC) $(ARCH_CFLAGS)+crypto -c sse2neon.h
```
Comment (jserv) [inline Makefile:74]: The use of `+crypto` looks ugly.

File: .github/workflows/github_actions.yml [hunk 4]
```diff
@@ -57,6 +57,7 @@ jobs:
       matrix:
         arch: [armv7, aarch64]
         cxx_compiler: [g++-10, clang++-11]
+        feature: ["none", "crypto"]
```
Comment (howjmay) [inline .github/workflows/github_actions.yml:60]: It seems `uraimo/run-on-arch-action@v2.5.0` doesn't allow us to set an empty element in the matrix, so I use a `none` element to represent the situation

File: Makefile (hunk 3 above)
Comment (howjmay) [reply Makefile]: I have removed this target

File: Makefile [hunk 5]
```diff
@@ -39,6 +39,12 @@ else
     $(error Unsupported architecture)
 endif
 
+FEATURE ?=
+ifneq ($(FEATURE),)
```
Comment (jserv) [inline Makefile:43]: We should take `make FEATURE=none` into consideration.

Comment (howjmay) [reply Makefile:43]: I have excluded `FEATURE =none` in GitHub Actions workflow. Or should I move the condition check to here?

Comment (jserv) [reply Makefile:43]: You don't have to exclude `FEATURE=none` in GitHub Actions since we always do `make check`. Just refine `Makefile` instead.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `master` branch (force-pushed earlier).

**Comment by howjmay** [comment]
A warning `WARNING: The requested image's platform (linux/arm/v7) does not match the detected host platform (linux/amd64) and no specific platform was requested` is returned, and I don't think the CI has been successfully run. I am investigating it

File: Makefile [hunk 6]
```diff
@@ -39,6 +39,14 @@ else
...
+ifneq ($(FEATURE),none)
+COMMA:= ,
+ARCH_CFLAGS := $(ARCH_CFLAGS)+$(subst $(COMMA),+,$(FEATURE))
```
Comment (jserv) [inline Makefile:46]: What if `make FEATURE=crypto,crc` or `make FEATURE=crypto+crc` is run?

Comment (howjmay) [reply Makefile:46]: Now the command should be `make FEATURE=crypto,crc check`. However, the bigger issue this PR is facing is that CI  is not successfully triggered on armv7

Comment (jserv) [reply Makefile:46]: Would we rollback older version of `run-on-arch-action`? Meanwhile, we should file a new issue there.

Comment (howjmay) [reply Makefile:46]: I think it was an error caused by I set the workflow yaml in wrong way. If not include this PR, the current CI seems to work fine

Comment (jserv) [reply Makefile:46]: By the way, you should modify `README.md` for the `FEATURE=` information while running `make check`.

Comment (howjmay) [reply Makefile:46]: Information has been added, and the tests work fine now. I made some silly error.

File: .github/workflows/github_actions.yml [hunk 7]
```diff
@@ -55,7 +55,7 @@ jobs:
     runs-on: ubuntu-20.04
     strategy:
       matrix:
-        arch: [armv7, aarch64]
+        arch_with_features: [{arch: armv7, feature: none}, {arch: aarch64, feature: none}, {arch: aarch64, feature: crypto}]
```
Comment (jserv) [inline .github/workflows/github_actions.yml:58]: I have changed the subject of this pull request and would like to address `crc` feature in CI pipeline as well.

Comment (howjmay) [reply .github/workflows/github_actions.yml:58]: Thanks! Btw should we remove the current crc enabling in Makefile line 33?

Comment (jserv) [reply .github/workflows/github_actions.yml:58]: Yes, go ahead.

Comment (howjmay) [reply .github/workflows/github_actions.yml]: done

**Pull request merged**